### PR TITLE
feat: BDD and evaluation framework (#101)

### PR DESCRIPTION
**Pull request by pkarw**
The idea is that the BDD valuation takes place after every iteration, which is a little bit suboptimal but, on the other hand, lets the users control the entire flow (and, for example, break it if something goes unexpected). 

Related to: #101

### CODE REVIEW

**Comment by pkarw** [comment]
I've got a problem with figuring out - on the telemetry level - that the agent finished the task and got its whole messages stacked with tool results. @grabbou could you please look at it and maybe suggest something? 
`snapshot`, which I'm using right now, is called in case of `tool` calls because of the `paused` intermediary state - but when there are no tools + when the tool result is processed, there are two edge cases I need to collect to be sure the per-agent test cases are properly scheduled.

**Comment by pkarw** [comment]
To test it run `bun run ./example/src/surprise_trip.test.ts`.

Expected behavior:

image

**Comment by pkarw** [comment]
OK @grabbou, the PR -> it's ready to rock :)

**Comment by grabbou** [comment]
From pictures, it is looking really good!

**Review by grabbou — APPROVED** [review]
This is amazing! Love it. I think the plan for tomorrow or early next week is to get this up and running on the CI on every commit, now that you have written tests for all our suites, and see how stable is it! 

I left some small comments, but nothing really that important. I will let you address them tomorrow, and then will proceed with merge and release. 

Ultimately, I would love to explore how we can make his workflow running via test runner, something similar/closer to things such as Vitest and Mocha. 

I am going to CC here @kraenhansen for second opinion and some thoughts, as this seems to be very much related to our another upcoming project, for which this could act as a great starting point!

**Comment by pkarw** [comment]
Thanks 🙏 I'll address the comments shortly

File: README.md [hunk 1]
```diff
@@ -426,6 +427,11 @@ If you want to handle tool execution manually, you can use `iterate` function to
 
 Have a look at how `teamwork` is implemented [here](./packages/framework/src/teamwork.ts) to understand how it works.
 
+
+### BDD Testing
+
+There's a packaged called `fabrice-ai/bdd` dedicated to unit testing - actually to Behavioral Driven Development. [Check the docs](./packages/bdd/README.md).
```
[the remark this answers is not in the document]
Comment (pkarw) [reply README.md:433]: Tbh, I don't have any cool idea at the moment. I believe BDD is pretty self-explanatory, so let's keep it like this and maybe change it after something cool pops up  😅

File: packages/bdd/rslib.config.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      format: 'esm',
+      bundle: false,
+      output: {
+        distPath: {
+          root: 'dist',
+        },
+      },
+    },
+    {
+      source: {
```
[the remark this answers is not in the document]
Comment (pkarw) [reply packages/bdd/rslib.config.ts]: Sure! fixed!

File: packages/bdd/README.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+      id: '2_exampleAgentTest',
+    }
+  }
+)
+```
+
+## Mocking tools
+
+You are able to very easily mock-up the tools used by the agents. For example: tools requesting user attention, or answers could be mocked using the LLM as answering machines - to keep the tests automatic.
+
```
[the remark this answers is not in the document]
Comment (pkarw) [reply packages/bdd/README.md:219]: Thanks!

File: packages/bdd/README.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+Having this in mind one could use the following commands to run:
+
+- Running tests:
+```ts
+$ tsx library_photo_to_website.test.ts
+```
+
+- Running workflow:
+```ts
+$ tsx library_photo_to_website.ts
```
[the remark this answers is not in the document]
Comment (pkarw) [reply packages/bdd/README.md:39]: You're right. It might have been misleading. As for now I've changed the naming convention to something like: `library_photo_to_website.config.ts` instead of `workflow`

**Comment by pkarw** [comment]
Hey @grabbou  - I've addressed the issues.  Thanks for the comments. I've honestly had no idea for any better name, so kept `bdd` ;) That is a good point on a naming convention. Changed: `example_name.workflow.ts` to `example.name.config.ts` so it's obvious it's not a runnable one. Now, it's ready to be merged.
